### PR TITLE
Similarity experiment

### PR DESCRIPTION
**Pull request by berggren**
Plaso events has a data_type attribute and based on this classification of event types, we can calculate a similarity score:
- Shingle (split) message field into set of “words”
- Compute Jaccard Distance of sets
- Small distance (neighbor) → similar event
- No/few neighbors → rare event

Minhashing approximation & LSH for performance

PoC implemented at this stage, need to be enabled in config.
Implements #499

### CODE REVIEW

File: timesketch/lib/datastores/elastic.py [hunk 1]
```diff
@@ -282,6 +285,51 @@ def search(self,
...
+            query_dsl: Dictionary containing Elasticsearch DSL query
+            indices: List of indices to query
+            return_fields: List of fields to return
+
+        Returns:
+            Generator of event documents in JSON format
+        """
+
+        if not query_filter.get(u'limit'):
+            query_filter[u'limit'] = 10000
```
Comment (Onager) [inline timesketch/lib/datastores/elastic.py]: Move this to a class constant or similar

File: timesketch/lib/experimental/similarity.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+    @staticmethod
+    def _calculate_score(lsh, minhash, total_num_events):
+        """Calculate a score based on Jaccard distance.
+
+        The score is calculated based on how many similar events that there are
+        for the event being scored. This is called neighbours and we simply
+        calculate how many neighbours the event has devided by the total events
+        in the LSH.
+
+        Returns: A float between 0 and 1.
```
Comment (Onager) [inline timesketch/lib/experimental/similarity.py]: +Args

File: timesketch/lib/experimental/similarity.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+
```
Comment (Onager) [inline timesketch/lib/experimental/similarity.py]: Remove extra blank line

File: timesketch/lib/experimental/similarity.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+        return lsh, minhashes
+
+    @staticmethod
+    def _calculate_score(lsh, minhash, total_num_events):
+        """Calculate a score based on Jaccard distance.
+
+        The score is calculated based on how many similar events that there are
+        for the event being scored. This is called neighbours and we simply
+        calculate how many neighbours the event has devided by the total events
```
Comment (Onager) [inline timesketch/lib/experimental/similarity.py]: divided

File: timesketch/lib/experimental/similarity.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+    def _get_config(self):
+        """Get config for supplied data_type.
+
+        Returns:
+            Dictionary with configuration parameters.
+        """
+        config_dict = self.CONFIG_REGISTRY.get(self._data_type)
+
+        # If there are no config for this data_type, use generic config and set
```
Comment (Onager) [inline timesketch/lib/experimental/similarity.py]: -are +is
-generic +default

File: timesketch/lib/experimental/similarity.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+        """Initializes a similarity scorer config.
+
+        Args:
+            index: Elasticsearch index name.
+            data_type: Name of the data_type.
+        """
+        self._index = index
+        self._data_type = data_type
+        for k, v in self._get_config().items():
+            setattr(self, k, v)
```
Comment (Onager) [inline timesketch/lib/experimental/similarity.py]: I don't love this config design - it's really hard to follow what configuration options there are, and where they're set.

File: timesketch/lib/experimental/similarity.py (hunk 3 above)
Comment (berggren) [reply timesketch/lib/experimental/similarity.py]: Done.

File: timesketch/lib/experimental/similarity.py (hunk 2 above)
Comment (berggren) [reply timesketch/lib/experimental/similarity.py]: Done.

File: timesketch/lib/experimental/similarity.py (hunk 4 above)
Comment (berggren) [reply timesketch/lib/experimental/similarity.py]: Done.

File: timesketch/lib/experimental/similarity.py (hunk 5 above)
Comment (berggren) [reply timesketch/lib/experimental/similarity.py]: Done.

File: timesketch/lib/experimental/similarity.py (hunk 6 above)
Comment (berggren) [reply timesketch/lib/experimental/similarity.py]: Yeah, I agree. The approach I take here is to have the user (developer for now, before we break out the config) to configure data_type configs in the dict, and then initialise the config objects attributes from that.

Will be refactored in the future, but keeping for now.

File: timesketch/lib/datastores/elastic.py (hunk 1 above)
Comment (berggren) [reply timesketch/lib/datastores/elastic.py]: Done.
